### PR TITLE
Improve sync thread logic

### PR DESCRIPTION
**Pull request by madurangasiriwardena**

### CODE REVIEW

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/pipeline/persist/Persistor.java [hunk 1]
```diff
@@ -121,12 +122,10 @@
...
                                 transactionResults.add(result);
                             } catch (SQLException e) {
-                                log.error("Error occurred while data sync.", e);
-                                log.error(e.getSQLState());
-                                try {
-                                    Thread.sleep(5000);
-                                } catch (InterruptedException e1) {
-                                    //
+                                if (e instanceof SQLIntegrityConstraintViolationException) {
+                                    //ignore. this will be recovered.
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/pipeline/persist/Persistor.java:126]: add debug log.

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/internal/SyncClientComponent.java [hunk 2]
```diff
@@ -96,9 +96,18 @@ protected void activate(ComponentContext context) {
     protected void deactivate(ComponentContext context) {
 
         if (syncService != null) {
-            List<SyncDataTask> syncDataTaskList = syncService.getSyncDataTaskList();
-            for (SyncDataTask syncDataTask : syncDataTaskList) {
-                syncDataTask.shutdown();
+            ScheduledExecutorService executor = syncService.getExecutor();
+            try {
+                executor.shutdown();
+                executor.awaitTermination(5, TimeUnit.SECONDS);
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/internal/SyncClientComponent.java]: configure forceful shutdown timeout

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/pipeline/DataSyncPipeline.java [hunk 3]
```diff
@@ -110,12 +110,11 @@ public boolean processBatch() throws SyncClientException {
                         targetConnection.rollback();
                         sourceConnection.rollback();
                     } catch (SQLException e) {
-                        batchComplete = false;
                         log.error("Error while rolling back sync transaction on table: " + pipelineConfiguration
                                 .getTableName(), e);
                     }
                 }
-                complete = batchComplete && transactionResults.isEmpty();
+                complete = transactionResults.isEmpty() || !batchProcessingSuccess;
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/pipeline/DataSyncPipeline.java]: remove loop here. change complete logic to check transactionResults sixe < batch size.

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/SyncDataTask.java [hunk 4]
```diff
@@ -30,48 +30,29 @@
...
-                        if (log.isDebugEnabled()) {
-                            log.debug("Batch processing for table: " + table + " completed. Sleeping the thread " +
-                                    "for: " + syncInterval + "ms.");
-                        }
-                        Thread.sleep(syncInterval);
-                    } catch (InterruptedException e) {
-                        throw new RuntimeException("Error occurred while attempting to sleep the thread: " +
-                                                   Thread.currentThread().getName());
-                    }
+            boolean complete =  false;
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/SyncDataTask.java]: add mechanism to gracefully shutdown after completing a batch

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/SyncDataTask.java [hunk 5]
```diff
@@ -48,30 +51,27 @@ public void run() {
...
-                    try {
-                        if (log.isDebugEnabled()) {
-                            log.debug("Batch processing for table: " + table + " completed. Sleeping the thread " +
-                                    "for: " + syncInterval + "ms.");
-                        }
-                        Thread.sleep(syncInterval);
-                    } catch (InterruptedException e) {
-                        throw new RuntimeException("Error occurred while attempting to sleep the thread: " +
-                                                   Thread.currentThread().getName());
+                dataSyncPipeline.process();
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/SyncDataTask.java:54]: add a mechanism to stop dataSyncPipeline process after completing a batch

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/pipeline/persist/Persistor.java [hunk 6]
```diff
@@ -121,12 +122,10 @@
...
+                                } else {
+                                    log.error("Error occurred while data sync. SQL state: " + e.getSQLState(), e);
                                 }
                                 TransactionResult result = new TransactionResult(entry, false, e);
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/pipeline/persist/Persistor.java:133]: if there is one failure, there is no need to proceed with processing the other elements in the batch.

**Review by omindu — APPROVED** [review]

**Pull request merged**